src/kde/kde.t.hh

```diff
@@ -0,0 +1,356 @@
+
+// clang-format off
+
+#ifndef kde_kde_t_hh
+#define kde_kde_t_hh
+
+#include "kde.hh"
+#include <iostream>
+#include <numeric>
+#include <math.h>
+
+namespace kde {
+
+/*!
+ * reconstruction 
+ * \brief
+ *
+ * Cartesian geometry reconstruction of a 1D distribution
+ *
+ * \param[in] distribution 
+ * \param[in] position
+ * \param[in] band_width
+ * \param[inout] final_distribution returned final function distribution
+ *
+ * Test of kde.
+ */ 
+template<>
+template<>
+std::vector<double> kde<kde_coordinates::CART>::reconstruction<1>(const
+        std::vector<double> &distribution, const
+        std::vector<std::array<double,3>> &position, const
+        std::vector<std::array<double,3>> &band_width, const bool
+        domain_decomposed) const{
+    const int64_t local_size = distribution.size();
+    Check(static_cast<int64_t>(position.size())==local_size);
+    Check(static_cast<int64_t>(band_width.size())==local_size);
+    
+    int64_t size = local_size;
+    int64_t global_lower_bound = 0;
+    int64_t global_upper_bound = local_size;
+    // minimize global values and only allocate them in DD problems
+    std::vector<double> global_contribution;
+    std::vector<double> global_x_position;
+    double global_max_x=-1e60;
+    double global_min_x=1e60;
+    for (int i = 0; i<local_size; i++){
+        global_max_x = std::max(global_max_x,position[i][0]);
+        global_min_x = std::min(global_min_x,position[i][0]);
+        Insist(!(band_width[i][0]<0.0),"bandwidth must be positive");
+    }
+    if(domain_decomposed){
+      // calculate global off sets and min/max values
+      int n_ranks = rtt_c4::nodes();
+      std::vector<int64_t> rank_size(rtt_c4::nodes(),0);
+      rank_size[rtt_c4::node()] = local_size;
+      rtt_c4::global_sum(rank_size.data(), n_ranks);
+      size = std::accumulate(rank_size.begin(),rank_size.end(),0.0);
+      std::vector<int64_t> accum_rank_size(rank_size);
+      std::partial_sum(rank_size.begin(), rank_size.end(),
+                   accum_rank_size.begin());
+      rtt_c4::global_max(global_max_x);
+      rtt_c4::global_min(global_min_x);
+
+      if(rtt_c4::node()>0){
+          global_lower_bound = accum_rank_size[rtt_c4::node()-1];
+          global_upper_bound = accum_rank_size[rtt_c4::node()];
+      }
+
+
+      // set up global arrays
+      global_contribution.resize(size,0.0);
+      global_x_position.resize(size,0.0);
+
+      // build up global positions
+      for (int i = 0; i<local_size; i++)
+          global_x_position[i+global_lower_bound] = position[i][0];
+
+      rtt_c4::global_sum(global_x_position.data(),size);
+
+    }
+
+    const double scale = std::accumulate(distribution.begin(), distribution.end(), 0.0);
+    std::vector<double> result(local_size,0.0);
+    std::vector<double> normal(local_size,0.0);
+
+
+    // accumulate weighted contribution from all kernels for conservation correction
```

Comment:
If this is a performance bottleneck (O(N^2) with local mesh size) you'll want to consider implementing a neighbor list here

Reply:
Yes, this is certainly not optimal. we can talk more about the best way to optimize this when we start running large data sets.